src/epics/helpers/ChatHandler.js

```diff
@@ -0,0 +1,203 @@
+import atob from 'atob';
+import btoa from 'btoa';
+import miniLock from '../../utils/miniLock';
+import { nowUTC } from '../../utils/time';
+import { Subject } from 'rxjs/Subject';
+import { Observable } from 'rxjs/Observable';
+import { ajax } from 'rxjs/observable/dom/ajax';
+import 'rxjs/add/operator/mergeMap';
+import 'rxjs/add/operator/map';
+import 'rxjs/add/operator/mapTo';
+import 'rxjs/add/operator/catch';
+import 'rxjs/add/operator/do';
+import 'rxjs/add/observable/of';
+import 'rxjs/add/observable/from';
+
+class ChatHandler {
+
+    constructor(wsUrl) {
+        this.wsUrl = wsUrl;
+        this.wsMessageSubject = null;
+        this.wsUserStatusSubject = null;
+    }
+
+    onWsClose = (event) => {
+        this.wsMessageSubject.error(new Error('The websocket connection was closed!'));
+    }
+
+    onWsError = (event) => {
+        this.wsMessageSubject.error(new Error('Error occured on websocket connection!'));
+    }
+
+    onWsMessage = (event) => {
+        const data = JSON.parse(event.data);
+        if (data && data.ephemeral && data.ephemeral.length && data.ephemeral.length > 0) {
+            Observable.from(data.ephemeral)
+                .mergeMap(ephemeral =>
+                    this.createDecryptEphemeralObservable({ ephemeral, mID: this.mID, secretKey: this.secretKey }))
+                .mergeMap(this.resolveIncomingMessageTypeObservable)
+                .catch(error => console.error('An error occured in ChatHandler', error))
+                .subscribe();
+        }
+    }
+
+    resolveIncomingMessageTypeObservable = (message) => {
+        if (message.tags.type === 'chatmessage') {
+
+            const reader = new FileReader();
+            const chatMessageObservable =
+                Observable.fromEvent(reader, 'loadend')
+                .pluck('target', 'result')
+                .map(JSON.parse)
+                .do(contents => {
+                    this.wsMessageSubject.next({
+                        fromUsername: message.tags.from,
+                        maybeSenderId: message.maybeSenderId,
+                        message: contents.msg
+                    })
+                })
+
+            reader.readAsText(message.fileBlob);
+            return chatMessageObservable;
+        } else {
```

Comment:
@Spetastian Can you turn this into an `else if`, and then handle `else` by throwing an error saying "Unrecognized data with tags ..."?

Reply:
That would be a good idea. What will happen when the error is thrown, how will it tell the user that something went wrong?

Reply:
Using `displayAlert` sounds good to me